test/extensions/filters/network/zookeeper_proxy/zookeeper_filter_test.cc

```diff
@@ -0,0 +1,709 @@
+  }
+
+  void addString(Buffer::OwnedImpl& buffer, const std::string& str) const {
+    buffer.writeBEInt<uint32_t>(str.length());
+    buffer.add(str);
+  }
+
+  void addStrings(Buffer::OwnedImpl& buffer, const std::vector<std::string>& watches) const {
+    buffer.writeBEInt<uint32_t>(watches.size());
+
+    for (const auto& watch : watches) {
+      addString(buffer, watch);
+    }
+  }
+
+  void expectSetDynamicMetadata(const std::map<std::string, std::string>& expected) {
+    EXPECT_CALL(filter_callbacks_.connection_, streamInfo())
+        .WillRepeatedly(ReturnRef(stream_info_));
+    EXPECT_CALL(stream_info_,
+                setDynamicMetadata("envoy.filters.network.zookeeper_proxy", MapEq(expected)));
+  }
+
+  void expectSetDynamicMetadata(const std::map<std::string, std::string>& first,
+                                const std::map<std::string, std::string>& second) {
+    EXPECT_CALL(filter_callbacks_.connection_, streamInfo())
+        .WillRepeatedly(ReturnRef(stream_info_));
+    EXPECT_CALL(stream_info_, setDynamicMetadata(_, _))
+        .WillOnce(Invoke([first](const std::string& key, const ProtobufWkt::Struct& obj) -> void {
+          EXPECT_STREQ(key.c_str(), "envoy.filters.network.zookeeper_proxy");
+          protoMapEq(obj, first);
+        }))
+        .WillOnce(Invoke([second](const std::string& key, const ProtobufWkt::Struct& obj) -> void {
+          EXPECT_STREQ(key.c_str(), "envoy.filters.network.zookeeper_proxy");
+          protoMapEq(obj, second);
+        }));
+  }
+
+  ZooKeeperFilterConfigSharedPtr config_;
+  std::unique_ptr<ZooKeeperFilter> filter_;
+  Stats::IsolatedStoreImpl scope_;
+  std::string stat_prefix_{"test.zookeeper"};
+  NiceMock<Network::MockReadFilterCallbacks> filter_callbacks_;
+  NiceMock<Envoy::StreamInfo::MockStreamInfo> stream_info_;
+};
+
+TEST_F(ZooKeeperFilterTest, Connect) {
+  initialize();
+
+  Buffer::InstancePtr data(new Buffer::OwnedImpl(encodeConnect()));
+
+  expectSetDynamicMetadata({{"opname", "connect"}}, {{"bytes", "32"}});
+
+  EXPECT_EQ(Envoy::Network::FilterStatus::Continue, filter_->onData(*data, false));
+  EXPECT_EQ(1UL, config_->stats().connect_rq_.value());
+  EXPECT_EQ(32UL, config_->stats().request_bytes_.value());
+  EXPECT_EQ(0UL, config_->stats().decoder_error_.value());
+}
+
+TEST_F(ZooKeeperFilterTest, ConnectReadonly) {
+  initialize();
+
+  Buffer::InstancePtr data(new Buffer::OwnedImpl(encodeConnect(true)));
+
+  expectSetDynamicMetadata({{"opname", "connect_readonly"}}, {{"bytes", "33"}});
+
+  EXPECT_EQ(Envoy::Network::FilterStatus::Continue, filter_->onData(*data, false));
+  EXPECT_EQ(0UL, config_->stats().connect_rq_.value());
+  EXPECT_EQ(1UL, config_->stats().connect_readonly_rq_.value());
+  EXPECT_EQ(33UL, config_->stats().request_bytes_.value());
+  EXPECT_EQ(0UL, config_->stats().decoder_error_.value());
+}
+
+TEST_F(ZooKeeperFilterTest, Fallback) {
+  initialize();
+
+  Buffer::InstancePtr data(new Buffer::OwnedImpl(encodeBadMessage()));
```

Comment:
i dont think you need to use a pointer here since you just need a buffer you can pass by ref